File: packages/tolk/test/parity/unary_sqrt_f16/main.ml
(* Parity case: sqrt on float16 — exercises half-precision intrinsic paths. *)

open Tolk_uop
module B = Program_spec_builder


let kernel () =
  let dt = Dtype.float16 in
  let ptr = dt in
  let b = B.create () in
  let p0 = B.emit b (Param { slot = 0; dtype = ptr }) in
  let p1 = B.emit b (Param { slot = 1; dtype = ptr }) in
  let c0 =
    B.emit b
      (Const { value = Const.int Dtype.int32 0; dtype = Dtype.int32 })
  in
  let idx0 =
    B.emit b (Index { ptr = p0; idxs = [ c0 ]; dtype = ptr })
  in
  let ld = B.emit b (Load { src = idx0; alt = None; gate = None; dtype = dt }) in
  let sq = B.emit b (Unary { op = `Sqrt; src = ld; dtype = dt }) in
  let idx1 =
    B.emit b (Index { ptr = p1; idxs = [ c0 ]; dtype = ptr })
  in
  let _ = B.emit b (Store { dst = idx1; value = sq; gate = None }) in
  B.finish b

let () = Helpers.dump_stage7_program ~out_dir:Sys.argv.(1) (kernel ())

File: packages/tolk/test/parity/unary_sqrt_f16/main.py
#!/usr/bin/env python3
"""Parity case: sqrt on float16 — exercises half-precision intrinsic paths."""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from helpers import dump_stage7_program  # noqa: E402

from tinygrad.uop.ops import UOp, Ops, KernelInfo  # noqa: E402
from tinygrad.dtype import dtypes  # noqa: E402


def kernel():
    sink = UOp(Ops.SINK, src=(), arg=KernelInfo())
    a = UOp.param(0, dtypes.half, shape=(1,))
    b = UOp.param(1, dtypes.half, shape=(1,))
    idx = UOp.cconst(0, dtypes.int)
    idx_a = a.index(idx)
    ld = UOp(Ops.LOAD, src=(idx_a,))
    sq = UOp(Ops.SQRT, src=(ld,))
    idx_b = b.index(idx)
    store = UOp(Ops.STORE, src=(idx_b, sq))
    return [sink, a, b, idx, idx_a, ld, sq, idx_b, store]


if __name__ == "__main__":
    dump_stage7_program(kernel(), os.path.dirname(os.path.abspath(__file__)))
